### PR TITLE
Optimize deepcopy

### PR DESCRIPTION
**Pull request by hackaugusto**
While investigating performance issues with the REST Proxy API I took a profile using `py-spy`, the profiler shows the cost of deepcopy is significant for the produce endpoint. This PR replaces the use of `copy.deepcopy` with a faster version based on `ujson`, at the same time replaces all usages of `json` with `ujson`.

Using stdlib's deepcopy:

profile

Using the ujson deepcopy:

profile2

The commit also contains a small timeit script to compare the performance of the two implementations.

### CODE REVIEW

File: karapace/utils.py [hunk 1]
```diff
@@ -20,11 +20,24 @@
...
 
 log = logging.getLogger("KarapaceUtils")
 NS_BLACKOUT_DURATION_SECONDS = 120
 
 
+def deepcopy(obj: Any) -> Any:
+    """Replacement for the standard library deepcopy.
+
+    The stdlib's copy.deepcopy is notoriously slow. After benchmarking the REST Proxy produce API
+    the deepcopy was identified as a bootleneck. Currenlty it is faster to encode the objects as
```
Comment (mble) [inline karapace/utils.py:33]: Small typo
```suggestion
    the deepcopy was identified as a bottleneck. Currently it is faster to encode the objects as
```

File: karapace/kafka_rest_apis/__init__.py [hunk 2]
```diff
@@ -292,7 +291,7 @@ def cluster_metadata(self, topics: Optional[List[str]] = None) -> dict:
                         content_type="application/json",
                         status=HTTPStatus.INTERNAL_SERVER_ERROR,
                     )
-        return copy.deepcopy(self._cluster_metadata)
+        return deepcopy(self._cluster_metadata)
```
Comment (tvainika) [inline karapace/kafka_rest_apis/__init__.py:294]: Does this need deepcopy at all? For me looks like `self._cluster_metadata` is assigned in this method with new value, and then returned.

**Review by tvainika — COMMENTED** [review]
Good catch & explanation.

I'd rather see `import ujson as json` to avoid gazillion of lines changed just for `s/json/ujson/`.

Also what comes to usage of deepcopy, could it be optimized even more, see inline comment? Other deepcopy changes are in the test code, so not relevant to benchmarking.

**Comment by hackaugusto** [comment]
> I'd rather see import ujson as json to avoid gazillion of lines changed just for s/json/ujson/.

Their APIs are similar but not a drop in replacement, for instance `ujson.dump` doesn't accept `fp` as a keyword argument, but only a position argument. Using `ujson` is more explicit this.

> Other deepcopy changes are in the test code, so not relevant to benchmarking.

True, I changed for consistency.

File: karapace/config.py [hunk 3]
```diff
@@ -112,13 +112,13 @@ def set_config_defaults(config: Config) -> Config:
...
-    config_path.write_text(json.dumps(custom_values))
+    config_path.write_text(ujson.dumps(custom_values))
 
 
 def read_config(config_handler: IO) -> Config:
     try:
-        config = json.load(config_handler)
-    except json.JSONDecodeError as ex:
+        config = ujson.load(config_handler)
+    except ValueError as ex:
```
Comment (jjaakola-aiven) [inline karapace/config.py:121]: Error class will get corrected in the next release of ujson.
https://github.com/ultrajson/ultrajson/issues/497

File: karapace/kafka_rest_apis/__init__.py (hunk 2 above)
Comment (jjaakola-aiven) [reply karapace/kafka_rest_apis/__init__.py:298]: Cluster metadata gets mutated, so copy is needed. For example here: https://github.com/aiven/karapace/blob/2578805875d115146675d79c50df9501c3cc5edd/karapace/kafka_rest_apis/__init__.py#L676

I think this the copy is ok here. Alternative would be to create the copies where mutation occurs.

**Review by jjaakola-aiven — APPROVED** [review]

**Comment by hackaugusto** [comment]
@jjaakola-aiven , the PR had conflicts and needed a rebase. A new review is needed.

**Review by jjaakola-aiven — CHANGES_REQUESTED** [review]
The `karapace/avro_compatibility` uses `json` and not `ujson`.

**Comment by hackaugusto** [comment]
> The karapace/avro_compatibility uses json and not ujson.

Yes, because ujson can not be used there, at least I didn't find in the API the position of the parser error to implement the same behavior.

**Review by jjaakola-aiven — APPROVED** [review]

File: karapace/kafka_rest_apis/__init__.py (hunk 2 above)
Comment (tvainika) [reply karapace/kafka_rest_apis/__init__.py]: Seems we're back here with #366

Comment (hackaugusto) [reply karapace/kafka_rest_apis/__init__.py]: I didn't review the code to check how the dict is used, is a shallow copy sufficient?

**Pull request merged**